<launch>
  <arg name="use_recovery_behaviors" default="true"/>
  <arg name="converter" default="$(optenv GM_PARAM_COSTMAP_CONVERTER true)"/>
  <arg name="costmap_converter_plugin" default="$(optenv GM_PARAM_COSTMAP_CONVERTER_PLUGIN costmap_converter::CostmapToPolygonsDBSMCCH)"/>
  <arg name="simple_exploration" default="$(optenv GM_PARAM_SIMPLE_EXPLORATION false)"/>
  <arg name="global_planning_freq" default="$(optenv GM_PARAM_GLOBAL_PLANNING_FREQ 1)"/>
  <arg name="feasibility_check_no_poses" default="$(optenv GM_PARAM_FEASIBILITY_CHECK_NO_POSES 5)"/>
  <arg name="debug" default="false"/>

  <arg name="debug_prefix" value="" unless="$(arg debug)"/> <!-- valgrind - -tool=callgrind -->
  <arg name="debug_prefix" value="xterm -e gdb -ex run --args" if="$(arg debug)"/>

  <!-- xterm -e gdb -ex run - -args -->
  
  <include file="$(find turtlebot_navigation)/launch/includes/velocity_smoother.launch.xml"/>
  <env name="ROSCONSOLE_CONFIG_FILE" value="$(find nav_configs)/config/custom_teb_rosconsole.conf"/>

  <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen" clear_params="true" launch-prefix="$(arg debug_prefix)">
      <rosparam file="$(find nav_configs)/config/common_costmap_params.yaml" command="load" ns="global_costmap"/>
      <rosparam file="$(find nav_configs)/config/common_costmap_params.yaml" command="load" ns="local_costmap"/>
      <rosparam file="$(find nav_configs)/config/local_costmap_params.yaml" command="load"/>
      <rosparam file="$(find nav_configs)/config/global_costmap_params.yaml" command="load"/>
      <rosparam file="$(find nav_configs)/config/local_planner_common_params.yaml" command="load" ns="TebLocalPlannerROS"/>
      <rosparam file="$(find nav_configs)/config/move_base_params.yaml" command="load"/>
      <rosparam file="$(find nav_configs)/config/teb_costmap_converter_params.yaml" command="load" if="$(arg converter)"/> 
      
      <param name="recovery_behavior_enabled" value="false" unless="$(arg use_recovery_behaviors)"/>

      <param name="base_local_planner" value="teb_local_planner/TebLocalPlannerROS" />
      <rosparam file="$(find nav_configs)/config/teb_local_planner_params.yaml" command="load"/>
      
      <param name="TebLocalPlannerROS/costmap_converter_plugin" value="$(arg costmap_converter_plugin)" if="$(arg converter)"/>
      <param name="TebLocalPlannerROS/simple_exploration" value="$(arg simple_exploration)"/>
      <param name="TebLocalPlannerROS/feasibility_check_no_poses" value="$(arg feasibility_check_no_poses)"/>
      <param name="planner_frequency" value="$(arg global_planning_freq)"/>

      <remap from="cmd_vel" to="navigation_velocity_smoother/raw_cmd_vel"/>
  </node>

</launch>
